src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessHelicopter.cpp

```diff
@@ -0,0 +1,143 @@
+		snprintf(buffer, sizeof(buffer), "CA_SP0_ANG%u", i);
+		_param_handles.swash_plate_servos[i].angle = param_find(buffer);
+		snprintf(buffer, sizeof(buffer), "CA_SP0_ARM_L%u", i);
+		_param_handles.swash_plate_servos[i].arm_length = param_find(buffer);
+	}
+
+	_param_handles.num_swash_plate_servos = param_find("CA_SP0_COUNT");
+
+	for (int i = 0; i < NUM_CURVE_POINTS; ++i) {
+		char buffer[17];
+		snprintf(buffer, sizeof(buffer), "CA_HELI_THR_C%u", i);
+		_param_handles.throttle_curve[i] = param_find(buffer);
+		snprintf(buffer, sizeof(buffer), "CA_HELI_PITCH_C%u", i);
+		_param_handles.pitch_curve[i] = param_find(buffer);
+	}
+
+	updateParams();
+}
+
+void ActuatorEffectivenessHelicopter::updateParams()
+{
+	ModuleParams::updateParams();
+
+	int32_t count = 0;
+
+	if (param_get(_param_handles.num_swash_plate_servos, &count) != 0) {
+		PX4_ERR("param_get failed");
+		return;
+	}
+
+	_geometry.num_swash_plate_servos = math::constrain((int)count, 3, NUM_SWASH_PLATE_SERVOS_MAX);
+
+	for (int i = 0; i < _geometry.num_swash_plate_servos; ++i) {
+		float angle_deg{};
+		param_get(_param_handles.swash_plate_servos[i].angle, &angle_deg);
+		_geometry.swash_plate_servos[i].angle = math::radians(angle_deg);
+		param_get(_param_handles.swash_plate_servos[i].arm_length, &_geometry.swash_plate_servos[i].arm_length);
+	}
+
+	for (int i = 0; i < NUM_CURVE_POINTS; ++i) {
+		param_get(_param_handles.throttle_curve[i], &_geometry.throttle_curve[i]);
+		param_get(_param_handles.pitch_curve[i], &_geometry.pitch_curve[i]);
+	}
+}
+
+bool
+ActuatorEffectivenessHelicopter::getEffectivenessMatrix(Configuration &configuration,
+		EffectivenessUpdateReason external_update)
+{
+	if (external_update == EffectivenessUpdateReason::NO_EXTERNAL_UPDATE) {
+		return false;
+	}
+
+	// As the allocation is non-linear, we use updateSetpoint() instead of the matrix
+	configuration.addActuator(ActuatorType::MOTORS, Vector3f{}, Vector3f{});
+
+	// Tail (yaw) servo
+	configuration.addActuator(ActuatorType::SERVOS, Vector3f{0.f, 0.f, 1.f}, Vector3f{});
+
+	// N swash plate servos
+	_first_swash_plate_servo_index = configuration.num_actuators_matrix[0];
+
+	for (int i = 0; i < _geometry.num_swash_plate_servos; ++i) {
+		configuration.addActuator(ActuatorType::SERVOS, Vector3f{}, Vector3f{});
+	}
+
+	return true;
+}
+
+void ActuatorEffectivenessHelicopter::updateSetpoint(const matrix::Vector<float, NUM_AXES> &control_sp,
+		int matrix_index, ActuatorVector &actuator_sp)
+{
+	// Find index to use for curves
+	float thrust_cmd = -control_sp(ControlAxis::THRUST_Z);
+	float num_intervals = NUM_CURVE_POINTS - 1;
+	// We access idx + 1 below, so max legal index is (size - 2)
+	int idx = math::constrain((int)(thrust_cmd * num_intervals), 0, NUM_CURVE_POINTS - 2);
+
+	// Local throttle curve gradient and offset
+	float tg = (_geometry.throttle_curve[idx + 1] - _geometry.throttle_curve[idx]) * num_intervals;
+	float to = (_geometry.throttle_curve[idx]) - (tg * idx / num_intervals);
```

Comment:
This doesn't really make sense. Aren't we dividing the each section into `num_intervals`?
Therefore shouldn't the gradient be : DeltaValue/NumIntervals, instead of multiplying the NumIntervals? 🤔 

```suggestion
	float tg = (_geometry.throttle_curve[idx + 1] - _geometry.throttle_curve[idx]) / num_intervals;
	float to = (_geometry.throttle_curve[idx]) - (tg * idx);
```

Reply:
Please check the math first if you're not sure. It checked out for me. And yes there's different ways to write this.